### ISSUE TITLE
Refactor type system

### ISSUE THREAD

**Issue by greg0ire** [doctrine/dbal#2841]
From @Majkl578 on slack:

just a quick thought of a possible approach:

    introduce TypeInterface with required methods (or different name)
    make types implement it
    introduce some kind of Types enum-like empty class for implictly supported types, just like i.e. GeneratorType in ORM develop (edit: maybe this could stay in Type to lower BC implications and make Type an enum class in 3.0)
    introduce non-static TypeRegistry with no default types hardcoded in it - this would allow different connections/platforms use different type mappings (as @Ocramius suggested in https://symfony-devs.slack.com/archives/C3FQPE6LE/p1504362019000073)
    make TypeRegistry work with type instances instead of their class names (sounds good, but unsure about consequences)
    migrate Type:: to Types enum, keep BC by aliasing them and deprecate in 2,7
    migrate Type:: mechanics to TypeRegistry, keep BC by using private static TypeRegistry internally in 2.7 (some kittens may die during this process)
    in 3.0 (develop) drop these BC hacks (so no more kittens die)

**Comment by greg0ire** [comment]
> use FQCNs instead of strings to identify types

In light of the need to have so-called parameterized types, this seems wrong. If we are going that way, the type registry should look more like a PSR-11 container (which it does already). FQCNs could still be used to identify types _by default_, but one should be able to have a type called `Doctrine\DBAL\Types\Enum.Country`.

**Comment by Ocramius** [comment]
TBH, just register `My\Enum\Type\Country`, or `My\Enum\Type<My\Domain\Thing\Country>`.

EDIT: to be clear, being able to have a `class-string<DBALType>` would already help some validation a lot, especially in attributes.

**Comment by greg0ire** [comment]
Oh yes, that second one sounds like a great idea, it would make it easier for static analysis tools to get make sense of the output of the type registry! So `Doctrine\DBAL\Types\Enum<MyProject\Country>`

**Comment by Ocramius** [comment]
Right, so:

```php
/** @template-extends SomeEnum<Country> */
class MyCountryEnum extends SomeEnum
{
    // ...
}
```

therefore you would register it as `MyCountryEnum::class` :D

**Comment by greg0ire** [comment]
Ah sorry, I should have made it a bit more obvious that I was referring to PHP 8 enums

It would look more like this:

```php
<?php
enum MyProject\Country
{
    // …
}

/**
 * @template E of enum-string
 */
class Doctrine\DBAL\Types\Enum
{
    /** @param E $enum */
    public function __construct(string $enum)
    {
        // ...
    }

    /** @return E */
    public function convertToPHPValue()
    {
    }
}
```

Then you'd be able to register `Doctrine\DBAL\Types\Enum<MyProject\Country>` without at any point creating a custom DBAL type that would know need to explicitly reference your enum. More on this at https://github.com/doctrine/dbal/pull/4930#issuecomment-953059917

**Comment by Ocramius** [comment]
IMO just subclass and provide `@template-extends`: reduces the problem, and you can use `class-string<DBALType>` directly, removing the whole custom naming problem.

I just had an exact problem around custom names today:

```php
/** @Column(type="some_custom_thing") */
```

Then proceed to have `"some_custom_thing"` mean different things depending on how your system is bootstrapped: total mess.

Better:


```php
/** @Column(type=MySpecificDomainType::class) */
```

This removes ambiguity, and allows using tools like https://github.com/qossmic/deptrac on your system, preventing usage of DBAL types unknown to the current namespace (where not allowed).

If it is static, declare it static, and use a class name. I would totally be for random strings if a PSR-11 container was used instead, but it isn't the case.

**Comment by greg0ire** [comment]
>  just subclass

If I have 10 enums, won't that mean 10 classes with very similar code? Won't that get old real fast?

> Then proceed to have "some_custom_thing" mean different things depending on how your system is bootstrapped: total mess.

Make sure you write your memoir someday :smile: 

> I would totally be for random strings if a PSR-11 container was used instead, but it isn't the case.

Well maybe this could be considered. We're not far from it:

https://github.com/doctrine/dbal/blob/187916074db413f4a070a721c9d30b420a9135dd/src/Types/TypeRegistry.php#L59-L65

https://github.com/doctrine/dbal/blob/187916074db413f4a070a721c9d30b420a9135dd/src/Types/TypeRegistry.php#L29-L41

**Comment by Ocramius** [comment]
> Won't that get old real fast?

No, usually just a class with no body suffices:

```php
/** @template-extends SomeEnumType<Country> */
final class MyCountryEnumType extends SomeEnumType
{
    // literally nothing here
}
```

**Comment by greg0ire** [comment]
Ok, so the code resides in `SomeEnumType`… how does that code figure out that when it reads `DE` in the database, it needs to return `Country::DE`? Would it read it from the annotation/attribute somehow?

**Comment by Ocramius** [comment]
Good point - indeed need to configure an enum reference or such :thinking: 

Probably:

```php
/** @template-extends SomeEnumType<Country> */
final class MyCountryEnumType extends SomeEnumType
{
    public function __construct() {
        parent::__construct(Country::class);
    }
}
```

**Comment by greg0ire** [comment]
Yes, and we're back to square one: types cannot be parameterized yet. Maybe I should open another RFC just for that.

**Comment by Ocramius** [comment]
@greg0ire that is parametrized? `MyCountryEnumType::class` is a `class-string<SomeEnumType<T of Enum>>`

**Comment by greg0ire** [comment]
No I meant `SomeEnumType`, but indeed `MyCountryEnumType` would work, and having an `abstract SomeEnumType` laying around would not hurt I suppose… there would be no need to register it.

**Comment by kadet1090** [comment]
This will still require us to write another (even if simple) class for every enum. That classes would be really repetitive, boring and provide no real value except from being workaround for not supporting type parametrization in DBAL / Doctrine as a whole.

Not to mention that this solution would be only really viable for cases when one parameter is required, like with enums. If for example I would want to have "encrypted" type, which can accept `cipher` parameter and `secretName` (which then would be used to obtain given secret from some kind of secure store service) parameters it'd force me to create class for every single parameter combination. And that does not seem fine for me.

**Comment by Ocramius** [comment]
> That classes would be really repetitive, boring and provide no real value except from being workaround for not supporting type parametrization in DBAL / Doctrine as a whole.

The type is parametrized: you register an actual instance of it. As for repetition: it's still a degree safer than what we had before, which is a registry of strings of unknown shape => shape is now known and well defined.

 > And that does not seem fine for me.

If the flyweight pattern is to be perpetuated around DBAL types, then there's no solution in sight for that anyway.

**Comment by greg0ire** [comment]
I wonder how important it is to perpetuate that pattern.

**Comment by Ocramius** [comment]
Probably not much, since PHP's memory usage is not such a huge problem as per PHP 7.0 (and this stuff was designed for 5.2.x and then 5.3.x).

The breakage would be ginormous, but having the `TypeRegistry` as runtime instance rather than singleton would suffice.

That opens the door for people to provide their own `TypeRegistry`, with blackjack and decorators.

**Comment by greg0ire** [comment]
> The breakage would be ginormous

@Ocramius @morozov @derrabus please have a look at https://github.com/doctrine/dbal/pull/5034 for an evaluation of said breakage. You will probably have better ideas than I did.

**Comment by greg0ire** [comment]
@Ocramius it's unclear to me how having the `TypeRegistry` as a runtime instance would get us rid of the flyweight pattern though… lifting that restriction seems like it would just require changing `Type::addType` / drop `Type::__construct()` and change `TypeRegistry::register()` to not throw when a type is already registered. Or did you mean that people would implement a `TypeRegistry` without that restriction and inject it?

It's also unclear to me what consequences lifting the "flyweight" restriction would have… do you see any there any?

I understand that restriction as follows:
- there can only be one instance of a type
- a type cannot have a state

**Comment by filiplikavcan** [comment]
Hi, I've created a PR which would allow to automatically register all child class (enum) type names by registering their parent class (enum) type name. Then it would be possible to implement one EnumType, register it for BackedEnum type name and it would be automatically used for any backed enum. See the example implementation here: https://github.com/doctrine/dbal/pull/5147

**Comment by tigitz** [comment]
@greg0ire Just to give some food for thought on this topic.

We use a specific type that deserialize json into a given object in the `convertToPHPValue` using `symfony\serializer`.

Similar to the mechanics of https://github.com/goodwix/doctrine-json-odm. 

What we would like to achieve from a consumer perspective is this:

```php
class Foo
       public function __construct(
        #[ORM\Column(type: new JsonSerializeType(Bar::class, serializationContext: [])]
        private Collection $bars,
    ) {
    }
```

Or even better

```php
class Foo
       public function __construct(
        /**
        * @param Collection<Bar>
        */
        #[ORM\Column(type: new JsonSerializeType(serializationContext: [])
        private Collection $bars,
    ) {
    }
```

That would require to have the `ReflectionProperty` of `Foo::$bar` passed to the Type somehow so that the type itself can read the docblock and proceed to the conversion accordingly.

Not sure if it's out of the scope of what has been described here and sorry it's not the case.

**Comment by greg0ire** [comment]
I think it's in scope. For me the next step towards that would be this: https://github.com/doctrine/dbal/pull/5036

**Comment by noemi-salaun** [comment]
@greg0ire 
Is there any plan to map native php enumerables to mysql's enum type? And is it in the scope of this issue?
It hurts to see a column of type varchar(255) when the enum is clearly defined and doctrine has all the info to create a perfect mysql enum :/

**Comment by greg0ire** [comment]
No plan AFAIK, and I don't think it's in scope of this issue, which is about providing a more powerful type system, not more powerful types.
Regarding enum's, I personally think they should be avoided: http://komlenic.com/244/8-reasons-why-mysqls-enum-data-type-is-evil/

**Comment by noemi-salaun** [comment]
Thanks for the quick answer 😄 

I see the debate. But isn't it overkill to use a varchar(255) to store values that rarely exceed 10 characters?
I can manualy set the length of my column with attribute/annotation for the ORM but I would rather just write

```php
#[Column]
private MyEnum $myEnum;
```

**Comment by greg0ire** [comment]
It is overkill yes: it takes more space than necessary.
